torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -134,7 +134,9 @@ std::vector<at::Device> getDeviceList(const std::vector<at::Tensor>& tensors) {
   std::vector<at::Device> res;
   res.reserve(tensors.size());
   for (auto& tensor : tensors) {
-    res.push_back(tensor.device());
+    if (res.size() == 0 || tensor.device() != res[0]) {
```

Comment:
Maybe I am too tired right now, but not sure if I follow the logic here. Is our goal to get the list of unique devices? If yes, how does this work for, say, `[cuda0, cuda1, cuda1]`? Looks like this logic will return `[cuda0, cuda1, cuda1]` instead of `[cuda0, cuda1]`.

Reply:
You're right, but the fact that tensors are either all on the same device, or all on different devices, is [enforced by check_gpu_tensors](https://github.com/pytorch/pytorch/pull/62140/files#diff-c4e79e1dee3eea72b89d4471b91de8a7dae284e27bd5c32ed3e5296a561e4ee9R1042-R1045).

Reply:
I see. Do you mind briefly mentioning this precondition as a comment here? Otherwise people will get confused about how this works.